Send metrics over GRPC even if the node is not in the `default` network (#1008)

mcamou opened the pull request.

willscott commented on metrics/metrics.go:
```diff
@@ -389,27 +389,71 @@ func GroupHandler() http.Handler {
+	}
 }
 
+// handlerForPeer returns the http.Handler associated with a given peer address.
+// If a peer belongs to multiple groups, it will return the handler associated with the
+// one that appears first in groupHandlers.
```
this seems weird - would we not prefer to extend the metrics protocol to allow peers to request a specific group that they want metrics back for?

mcamou replied:
I see what you mean in that it feels a bit weird, however,the metrics are group-agnostic (they contain metrics for all groups that the node is a part of). If we made it so that you could request for a specific group, we would then have to modify deployment in `drand-infra` each time we add a new beacon. What we want here is to be able to get the metrics for all nodes that this node is connected to, regardless of the group.

We could refactor it so that we have a global, non-group-dependent GRPC connection for the sole purpose of sending metrics (and possibly other control messages in the future), but I'm not sure that that would be worth it at this point.

mcamou replied:
Another option I thought of over the weekend, is to populate a `sync.Map` whenever we try to connect to a host. I believe that would be here: https://github.com/drand/drand/blob/master/net/client_grpc.go#L325. We could add something like this after the `if err != nil { ... }` block, assuming that `metrics.Handlers` is a `sync.Map`:

```go
metrics.Handlers.Store(p.Address(), g.HandleHTTP(p)
```

we would then entirely remove`MetricsHandlerForPeer` and `handlerForPeer`, and replace the call to `handlerForPeer` here https://github.com/drand/drand/pull/1017/files#diff-d715f391f2ffa038c3167bfea1adf741c1f5e39f08349fbbacab780a594146a4R453-R460 with:

```go
handler, found := Handlers.Load(addr)
if !found {
  log.DefaultLogger().Warnw("", "metrics", "peer not found", "addr", addr)
  w.WriteHeader(http.StatusNotFound)
  return
}
```

It does feel much cleaner that way, WDYT?

mcamou commented on metrics/metrics.go:
```diff
@@ -363,8 +363,8 @@ func Start(metricsBind string, pprof http.Handler, peerHandler PeerHandler) net.
-	if peerHandler != nil {
-		mux.Handle("/peer/", &lazyPeerHandler{peerHandler})
+	if groupHandlers != nil {
+		mux.Handle("/peer/", newLazyPeerHandler(groupHandlers))
```
I renamed peer -> group in those cases that affect a whole group (therefore e.g. `PeerHandler` -> `GroupHandler`, and in this case, `groupHandlers` (plural) contains handlers for each of the groups. However, the `lazyPeerHandler` does not handle a group, it handles all the peers that this host is connected to regardless of group.

AnomalRoil commented on core/drand_beacon_metrics.go:
```diff
@@ -1,28 +1,32 @@
 	hc, ok := pc.(net.HTTPClient)
 	if !ok {
-		return nil, errors.New("implementation does not support metrics")
+		return nil, errors.New("ProtocolClient implementation does not support metrics")
```
You'll need a `// nolint:golint` of start with a small letter to pass the linter

mcamou replied:
👍🏻

AnomalRoil commented on core/drand_beacon_metrics.go:
```diff
@@ -1,28 +1,32 @@
 	handlers := make(map[string]http.Handler)
 	var err error
 	for _, n := range bp.group.Nodes {
+		log.DefaultLogger().Infow("", "metrics", "adding node to metrics", "beaconID", bp.beaconID, "address", n.Address())
```
Please use the `bp.log` logger, so we know which `bp` emits the log.  
It already includes the `beaconID`.

mcamou replied:
👍🏻

AnomalRoil commented on core/drand_daemon.go:
```diff
@@ -248,18 +248,25 @@ func (dd *DrandDaemon) LoadBeaconsFromDisk(metricsFlag string) error {
-		// Start metrics server
 		if metricsFlag != "" {
-			_ = metrics.Start(metricsFlag, pprof.WithProfile(), bp.PeerMetrics)
+			log.DefaultLogger().Infow("", "metrics", "adding handler", "beacon", bp.beaconID)
```
Please us the `bp.log` logger, it already includes the bp.beaconID, btw.

mcamou replied:
👍🏻

AnomalRoil commented on metrics/metrics.go:
```diff
@@ -389,27 +389,74 @@ func GroupHandler() http.Handler {
+		}
+	}
+
+	return nil, err
```
This should probably return an error saying no handler was found, no?

mcamou replied:
good call

AnomalRoil commented on metrics/metrics.go:
```diff
@@ -389,27 +389,74 @@ func GroupHandler() http.Handler {
 	if index := strings.Index(addr, "/"); index != -1 {
 		addr = addr[:index]
 	}
+	log.DefaultLogger().Debugw("", "metrics", "getting metrics for peer", "addr", addr)
```
The `DefaultLogger` logs only at the `Info` level, so `Debug` level logs are no-ops.
If you want to support `Debug` level ops, you should add a `Logger` to your struct and have it be inherited from a `bp` or config logger when you initialize it.

AnomalRoil reviewed (changes requested):
Could you check my comments?

Also did you see the linter complaints around  `l := Start(":0", nil, mph)` in [metrics_test.go](/metrics/metrics_test.go)?
> cannot use mph (variable of type func(ctx context.Context) (map[string]http.Handler, error)) as []GroupHandlers value in argument to Start (typecheck)

mcamou commented:
@AnomalRoil I've addressed your comments (and done quite a bit of refactoring). Could you have another look?

@willscott I've done quite a bit of refactoring, but my replies to your comments above still stand. WDYT?

AnomalRoil commented on metrics/metrics.go:
```diff
@@ -389,27 +389,79 @@ func GroupHandler() http.Handler {
 
-	handler, ok := handlers[addr]
-	if !ok {
+	if handler == nil {
```
This can never happen when `err == nil`, can it?

mcamou replied:
See https://github.com/drand/drand/pull/1017/files#diff-d715f391f2ffa038c3167bfea1adf741c1f5e39f08349fbbacab780a594146a4R444. `nil, nil` is returned when there is no handler. I'll use the same pattern as for `ErrNoGroup` to make it clearer.

mcamou commented on common/beacon.go:
```diff
@@ -41,3 +43,12 @@ func GetCanonicalBeaconID(id string) string {
 }
+
+// NotPartOfGroupError indicates that this node is not part of the group for a specific beacon ID
+type NotPartOfGroupError struct {
```
Thanks... I'm a newbie with Go coding style and keep falling back on patterns from other languages.

mcamou commented on core/drand_beacon_metrics.go:
```diff
@@ -1,34 +1,40 @@
+				return h, nil
+			}
+
+			bp.log.Infow("", "metrics", "Error while adding node", "address", n.Address(), "error", err)
```
👍🏻

mcamou commented on metrics/metrics.go:
```diff
@@ -389,27 +389,79 @@ func GroupHandler() http.Handler {
+
+	var err error
+
+	for _, hdl := range l.metricsHandlers {
```
👍🏻

mcamou commented on metrics/metrics.go:
```diff
@@ -389,27 +389,79 @@ func GroupHandler() http.Handler {
+		h, err = hdl(addr)
+		if err != nil {
+			//nolint:errorlint
+			switch err.(type) {
```
👍🏻

mcamou commented on metrics/metrics.go:
```diff
@@ -363,8 +363,8 @@ func Start(metricsBind string, pprof http.Handler, peerHandler PeerHandler) net.
 	mux := http.NewServeMux()
 	mux.Handle("/metrics", promhttp.HandlerFor(PrivateMetrics, promhttp.HandlerOpts{Registry: PrivateMetrics}))
```
From looking at the code, `metrics.Start` is only called when the `--metrics` flag is set, so it should not be an issue.

AnomalRoil commented:
LGTM now but I'd love to have Will do another pass on it before merging, since I never touched the metrics so far.

willscott reviewed (approved):
I'm okay with this.

relaying of peer metrics was initially done as a very light-weight way to get some views into other group members.

Because of the need for manual addressing of all the collected endpoints, i do wonder if we've exhausted it's usefulness, and should think about what we actually want as a way to aggregate metrics across peers and groups.
in particular, it would be useful potentially to have a single endpoint, and to put thought into managing the cardinality of metrics down to what we actually want

mcamou commented:
@AnomalRoil with Will's 👍🏻, could you please approve?
